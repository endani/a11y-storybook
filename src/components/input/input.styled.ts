import styled from "styled-components"
import { InputProps } from "./input.types"

export const InputWrapper = styled.input<InputProps>`
  font-family: "Nunito Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
  background: none;
  border: none;
  border-bottom: 1px solid #515151;
  color: #515151;
  font-size: 16px;
  font-style: normal;
  font-weight: 400;
  letter-spacing: -0.0125em;
  line-height: 140%;
  outline: none;
  padding: 8px 2px;
  width: 100%;

  ::placeholder {
    color: #515151;
    opacity: 0.2;
  }
`
