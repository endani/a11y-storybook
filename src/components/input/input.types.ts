import { ChangeEvent, RefObject } from "react"

export interface InputProps {
  autoComplete?: string
  autoFocus?: boolean
  id?: string
  inputRef?: RefObject<HTMLInputElement>
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void
  placeholder: string
  role?: string
  tabIndex?: number
  type: string
  value?: string
}
