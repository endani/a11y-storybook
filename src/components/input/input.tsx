import { debounce } from "lodash"
import { InputWrapper } from "./input.styled"
import { InputProps } from "./input.types"

const INPUT_DEBOUNCE_TIME = 100

const Input = ({ onChange, inputRef, ...props }: InputProps) => {
  return (
    <InputWrapper
      ref={inputRef}
      data-qa="input"
      onChange={onChange && debounce(onChange, INPUT_DEBOUNCE_TIME)}
      {...props}
    />
  )
}

export { Input }
