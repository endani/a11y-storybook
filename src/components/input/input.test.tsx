import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"

import { Input } from "./input"
import { InputProps } from "./input.types"

const mockOnChange = jest.fn()

jest.mock("lodash", () => ({
  debounce: (fn: any) => fn,
}))

const renderInput = ({ placeholder, type, ...props }: InputProps) =>
  render(
    <Input
      {...props}
      placeholder={placeholder}
      type={type}
      onChange={mockOnChange}
    />
  )

describe("Input", () => {
  it("should render a Input correctly", () => {
    const { container } = renderInput({
      placeholder: "Awesome Input",
      type: "text",
    })

    expect(container).toMatchSnapshot()
    expect(screen.getByTestId("input")).toBeInTheDocument()
  })

  it("should render a search Input correctly", () => {
    renderInput({
      placeholder: "Awesome Input",
      type: "search",
    })

    expect(screen.getByTestId("input")).toHaveAttribute("type", "search")
  })

  it("should render a Input with a custom placeholder", () => {
    renderInput({
      placeholder: "Nice Input",
      type: "text",
    })

    expect(screen.getByTestId("input")).toHaveAttribute(
      "placeholder",
      "Nice Input"
    )
  })

  it("should call onChange when the input value changes", async () => {
    renderInput({
      placeholder: "Awesome Input",
      type: "text",
    })

    const input = screen.getByTestId("input")

    await userEvent.type(input, "New Value")

    expect(mockOnChange).toBeCalled()

    expect(mockOnChange).toBeCalledTimes(9)
  })
})
