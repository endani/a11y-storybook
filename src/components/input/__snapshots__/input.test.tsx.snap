// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render a Input correctly 1`] = `
<div>
  <input
    class="sc-aYaIB iXnLvK"
    data-qa="input"
    placeholder="Awesome Input"
    type="text"
  />
</div>
`;
